src/environment/providers/location.ts

```diff
@@ -22,6 +22,33 @@ export class TextDocumentLocationProviderRegistry<
     public getLocation(params: P): Observable<L | L[] | null> {
         return getLocation<P, L>(this.providers, params)
     }
+
+    public getLocationsWithProviderName(params: P): Observable<{ providerName: string; location: L }[] | null> {
+        return getLocationsWithProviderName<P, L>(this.providersWithName, params)
+    }
+
+    public readonly providersWithName: Observable<
+        { name: string; provider: ProvideTextDocumentLocationSignature<P, L> }[]
+    > = this.entries.pipe(
+        map(providers =>
+            providers.map(({ provider, registrationOptions }) => ({
+                name: registrationOptions.providerName,
+                provider,
+            }))
+        )
+    )
+
+    public registerProvider(
+        registrationOptions: TextDocumentRegistrationOptions,
+        provider: ProvideTextDocumentLocationSignature<P, L>
+    ): Unsubscribable {
+        if (!registrationOptions.providerName) {
```

Comment:
It would be nice if there were a way to enforce this with the type system somewhere at a higher level than in this method body. Right now, the way you have it, you have some values of type `TextDocumentRegistrationOptions` that have a null or undefined `providerName` property. That means the type is lying, which is generally not good. Ideally you can make it so that the validation and cast occurs in the same place and the value never has a type that is incorrect.

Reply:
I added the abstract `validateRegistrationOptions` method to `Feature` and added implementations for all concrete subclasses. This is necessary, because we need to assert the validity of the options object when deserializing the registration request. This seemed like a reasonably clean way to do it that keeps the validating logic in the class that needs the type safety.